activities.el

```diff
@@ -651,6 +652,23 @@ activity's name is NAME."
     (run-at-time nil nil (lambda ()
                            (window-state-put state (frame-root-window) 'safe)))))
 
+(defun activities--map-window-state-leafs (state func)
+  "Call FUNC on all the leaf nodes in window-state STATE.
+The resulting values are returned as a list."
+  (let (ret)
+    (cl-labels ((map-leafs (state func)
+		  (pcase state
+		    (`(leaf . ,_attrs)
+		     (push (funcall func state) ret))
+		    ((pred atom) state)
+		    (`(,_key . ,(pred atom)) state)
+		    ((pred list)
+		     (if-let ((leaf-pos (cl-position 'leaf state)))
+			 (push (funcall func (cl-subseq state leaf-pos)) ret)
+		       (dolist (s state) (map-leafs s func)))))))
+      (map-leafs state func))
+    (nreverse ret)))
```

Comment:
You mean branches in the `pcase`, yes?  One of those is necessary to "do nothing" on "uninteresting lists" that may be present.  

BTW, this revealed a bug in your `pcase` formulation (which I borrowed from `bufferize-window-state`): `(pred list)` is always true.  I think `list` is not the predicate you need; I used `proper-list-p`.

Reply:
You're right that it was a mistake: `list` is not a predicate.  `listp` is what I should have used.  It may be that it made no difference in practice.  

I think using `proper-list-p` probably isn't necessary over `listp`; the values in the window state object are well-defined, and misusing `list` was apparently working.

Reply:
Oh, I see how you removed the other clause, so now using `proper-list-p` is likely necessary.

Reply:
It made no difference I think because all other cases are caught by your prior cases, so `t` would have even sufficed.   Probably best for safety to use `listp`.  I'm using `proper-list-p` because I need to exclude cons cells like `(min-width . 8)` which I no longer match affirmatively.  I could restore that "do-nothing" case or just insist on a "real list"; happy with whatever would make more sense to you down the road.  `cl-position` does not take kindly to non-nil-terminated cons cells.